contracts/modules/capital/SwapOperator.sol

```diff
@@ -347,4 +361,162 @@ contract CowSwapOperator {
+
+    // Validate there's no current cow swap order going on
+    require(currentOrderUID.length == 0, "SwapOp: an order is already in place");
+
+    IPool pool = _pool();
+    IERC20Detailed fromToken = IERC20Detailed(enzymeV4VaultProxyAddress);
+
+    uint balanceBefore = fromToken.balanceOf(address(pool));
+    {
+
+      IPool.SwapDetails memory swapDetails = pool.getAssetSwapDetails(address(fromToken));
+
+      require(!(swapDetails.minAmount == 0 && swapDetails.maxAmount == 0), "SwapOp: asset is not enabled");
+
+      // swap frequency check
+      uint timeSinceLastTrade = block.timestamp - uint(swapDetails.lastSwapTime);
+      require(timeSinceLastTrade > MIN_TIME_BETWEEN_ORDERS, "SwapOp: too fast");
+
+      uint netShareValue;
+      {
+        address denominationAsset;
+        (denominationAsset, netShareValue) =
+        enzymeFundValueCalculatorRouter.calcNetShareValue(enzymeV4VaultProxyAddress);
+
+        require(denominationAsset ==  address(weth), "SwapOp: invalid denomination asset");
+      }
+
+      // avgAmountOut in ETH
+      uint avgAmountOut = amountIn * netShareValue / (10 ** fromToken.decimals());
+      uint maxSlippageAmount = avgAmountOut * swapDetails.maxSlippageRatio / 1e18;
+      uint minOutOnMaxSlippage = avgAmountOut - maxSlippageAmount;
+
+      // slippage check
+      require(amountOutMin >= minOutOnMaxSlippage, "SwapOp: amountOutMin < minOutOnMaxSlippage");
+      require(balanceBefore > swapDetails.maxAmount, "SwapOp: balanceBefore <= max");
+      require(balanceBefore - amountIn >= swapDetails.minAmount, "SwapOp: tokenBalanceAfter < min");
+    }
+
+    pool.transferAssetToSwapOperator(address(fromToken), amountIn);
+
+    IEnzymeV4Comptroller comptrollerProxy = IEnzymeV4Comptroller(IEnzymeV4Vault(enzymeV4VaultProxyAddress).getAccessor());
+    fromToken.approve(address(comptrollerProxy), amountIn);
+
+    address[] memory payoutAssets = new address[](1);
+    uint[] memory payoutAssetsPercentages = new uint[](1);
+
+    payoutAssets[0] = address(weth);
+    payoutAssetsPercentages[0] = 10000;
+
+    comptrollerProxy.redeemSharesForSpecificAssets(address(this), amountIn, payoutAssets, payoutAssetsPercentages);
+
+    uint amountOut = weth.balanceOf(address(this));
+    weth.withdraw(amountOut);
+
+    pool.setSwapDetailsLastSwapTime(address(fromToken), uint32(block.timestamp));
+
+    require(amountOut >= amountOutMin, "SwapOp: amountOut < amountOutMin");
+
+    transferAssetTo(ETH, address(pool), amountOut);
+
+    emit Swapped(enzymeV4VaultProxyAddress, ETH, amountIn, amountOut);
+  }
+
+  function transferAssetTo (address asset, address to, uint amount) internal {
+
+    if (asset == ETH) {
+      (bool ok, /* data */) = to.call{ value: amount }("");
+      require(ok, "SwapOp: Eth transfer failed");
+      return;
+    }
+
+    IERC20 token = IERC20(asset);
+    token.safeTransfer(to, amount);
+  }
+
+  function recoverAsset(address assetAddress, address receiver) public onlyController {
```

Comment:
The `recoverAsset` function allows the swap controller to drain assets from this contract while there's an open limit order. We should add a check similar to the one in `swapEnzymeVaultShareForETH`.

Reply:
There's also the flip side of the coin: if we ever get stuck in the "limit order open" status - we won't be able to recover any stuck assets here. TODO: We need to make sure it is always possible to get out of an open order!

Reply:
added that check

Reply:
created a ticket for the TODO:

https://github.com/NexusMutual/smart-contracts/issues/450